mep/accounts/models.py

```diff
@@ -619,18 +600,85 @@ def date_range(self):
 
         # if both dates are set and the same, return a single date
         if self.partial_start_date and self.partial_end_date and \
-            self.partial_start_date == self.partial_end_date:
+                self.partial_start_date == self.partial_end_date:
             return self.partial_start_date
 
         # otherwise, use both dates with ?? to indicate unknown date
         return '/'.join([dt if dt else '??'
-                         for dt in [self.partial_start_date, self.partial_end_date]])
+                         for dt in [self.partial_start_date,
+                                    self.partial_end_date]])
+
+
+class Borrow(PartialDateMixin, Event):
+    '''Inherited table indicating borrow events'''
+    #: :class:`~mep.books.models.Item` that was borrowed;
+    #: optional to account for unclear titles
+    item = models.ForeignKey(Item, null=True, blank=True)
+    ITEM_RETURNED = 'R'
+    ITEM_BOUGHT = 'B'
+    ITEM_MISSING = 'M'
+    STATUS_CHOICES = (
+        ('', 'Unknown'),
+        (ITEM_RETURNED, 'Returned'),
+        (ITEM_BOUGHT, 'Bought'),
+        (ITEM_MISSING, 'Missing'),
+    )
+    item_status = models.CharField(max_length=2, blank=True,
+        help_text='Status of borrowed item (bought, missing, returned)',
+        choices=STATUS_CHOICES)
+    footnotes = GenericRelation(Footnote)
+
+    def save(self, *args, **kwargs):
+        # if end date is set and item status is not, automatically set
+        # status to returned
+        if self.end_date and not self.item_status:
+            self.item_status = self.ITEM_RETURNED
+        super(Borrow, self).save(*args, **kwargs)
 
 
-class Purchase(Event, CurrencyMixin):
-    '''Inherited table indicating purchase events'''
+class Purchase(PartialDateMixin, CurrencyMixin, Event):
+    '''Inherited table indicating purchase events; extends :class:`Event`'''
     price = models.DecimalField(max_digits=8, decimal_places=2)
     item = models.ForeignKey(Item)
+    footnotes = GenericRelation(Footnote)
+
+    def date(self):
+        '''alias of :attr:`date_range` for display; since reimbersument
+        is a single-day event will always be a single partial date.'''
+        return self.date_range
+    date.admin_order_field = 'start_date'
+
+    def save(self, *args, **kwargs):
+        # override save to always set start = end, end will be disabled in
+        # admin
+        self.end_date_precision = self.start_date_precision
+        self.end_date = self.start_date
+        super().save(*args, **kwargs)
+
+    def validate_unique(self, *args, **kwargs):
+        '''Validation check to prevent duplicate purchase events from happening.
+        Differs from
+        :class:`~mep.accounts.models.Reimbursement.validate_unique` by also
+        allowing for multiple purchases of different items per day.
+        '''
```

Comment:
Possible worth noting that we can't use `unique_together` because of multi-table inheritance. (I know it's noted on the other, but I forgot and had to look at it to see why we did this.)